geo/src/algorithm/kmeans/mod.rs

```diff
@@ -359,24 +370,51 @@ where
     let mut upper_bounds = vec![T::infinity(); n];
     let mut lower_bounds = vec![T::zero(); n];
 
-    // First iteration: assign all points and initialize bounds
-    for (i, ((point, assignment), (upper, lower))) in points
+    // First iteration: assign all points to initial (k-means++) centroids
+    for (i, (point, assignment)) in points.iter().zip(assignments.iter_mut()).enumerate() {
+        let (nearest_idx, _, _) = find_nearest_and_second_nearest(
+            *point,
+            &centroids,
+            point_sq_norms[i],
+            &centroid_sq_norms,
+        );
+        *assignment = nearest_idx;
+    }
+
+    // Update centroids to be cluster means (initial centroids from k-means++ are data points,
+    // not true cluster means). This is required before the first convergence check because
+    // otherwise points at the initial centroid locations have upper_bound=0 and get pruned.
```

Comment:
Where/why would they get pruned?

(Sorry I'm not very familiar with this algo yet...)